applications/GeoMechanicsApplication/custom_elements/transient_Pw_line_element.h

```diff
@@ -354,31 +293,110 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) TransientPwLineElement : public Elem
+    {
+        switch (rContribution) {
+        case CalculationContribution::Permeability:
+            return std::make_unique<PermeabilityCalculator>(CreatePermeabilityInputProvider());
+        case CalculationContribution::Compressibility:
+            return std::make_unique<CompressibilityCalculator>(CreateCompressibilityInputProvider(rCurrentProcessInfo));
+        default:
+            KRATOS_ERROR << "Unknown contribution" << std::endl;
+        }
+    }
+
+    CompressibilityCalculator::InputProvider CreateCompressibilityInputProvider(const ProcessInfo& rCurrentProcessInfo)
+    {
+        return CompressibilityCalculator::InputProvider(
+            GetPropertiesLambda(), GetRetentionLawsLambda(), GetNContainerLambda(),
+            GetIntegrationCoefficientsLambda(), GetDtPressureCoefficientLambda(rCurrentProcessInfo),
+            GetNodalVariableLambda());
+    }
+
+    PermeabilityCalculator::InputProvider CreatePermeabilityInputProvider()
+    {
+        return PermeabilityCalculator::InputProvider(
+            GetPropertiesLambda(), GetRetentionLawsLambda(), GetIntegrationCoefficientsLambda(),
+            GetNodalVariableLambda(), GetShapeFunctionLocalGradientsLambda());
+    }
+
+    auto GetPropertiesLambda()
+    {
+        return [this]() -> const Properties& { return GetProperties(); };
+    }
+
+    auto GetRetentionLawsLambda()
+    {
+        return [this]() -> const std::vector<RetentionLaw::Pointer>& { return mRetentionLawVector; };
+    }
+
+    auto GetNContainerLambda()
+    {
+        return [this]() -> const Matrix& {
+            return GetGeometry().ShapeFunctionsValues(GetIntegrationMethod());
+        };
+    }
+
+    auto GetIntegrationCoefficientsLambda()
+    {
+        return [this]() -> Vector {
+            Vector det_J_container;
+            GetGeometry().DeterminantOfJacobian(det_J_container, this->GetIntegrationMethod());
+            return CalculateIntegrationCoefficients(det_J_container);
+        };
+    }
+
+    static auto GetDtPressureCoefficientLambda(const ProcessInfo& rCurrentProcessInfo)
+    {
+        return [&rCurrentProcessInfo]() { return rCurrentProcessInfo[DT_PRESSURE_COEFFICIENT]; };
+    }
+
+    auto GetNodalVariableLambda() const
+    {
+        return
+            [this](const Variable<double>& variable) -> Vector { return GetNodalValuesOf(variable); };
+    }
+
+    auto GetShapeFunctionLocalGradientsLambda()
+    {
+        return [this]() {
+            Vector det_J_container;
+            GetGeometry().DeterminantOfJacobian(det_J_container, this->GetIntegrationMethod());
+            GeometryType::ShapeFunctionsGradientsType dN_dX_container =
+                GetGeometry().ShapeFunctionsLocalGradients(this->GetIntegrationMethod());
+            std::transform(dN_dX_container.begin(), dN_dX_container.end(), det_J_container.begin(),
+                           dN_dX_container.begin(), std::divides<>());
+
+            return dN_dX_container;
+        };
+    }
```

Comment:
GetShapeFunctionsLocalGradients the body contents seem to generic to be part of 1 specific element.

Reply:
Yes, I think this can be a utility function, is it an idea to create that in a separate PR?

Reply:
Yes.